src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,136 @@
+            add(value);
+        } else {
+            ensureCapacity();
+            checkAddPosition(index);
+            Object[] temp = new Object[storage.length];
+            System.arraycopy(storage, 0, temp, 0, index);
+            System.arraycopy(storage, index, temp, index + 1, size - index);
+            storage = temp;
+            storage[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int requiredSize = size + list.size();
+        if (requiredSize > storage.length) {
+            changeCapacity(requiredSize);
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkGetPosition(index);
+        return (T) storage[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkGetPosition(index);
+        storage[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkGetPosition(index);
+        T obj = (T) storage[index];
+        removeBy(index);
+        return obj;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T obj;
+        for (int i = 0; i < size; i++) {
+            obj = (T) storage[i];
+            if (obj == element || obj != null && obj.equals(element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException(NO_SUCH_ELEMENT_MESSAGE + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void optimizeSize() {
+        if (size < storage.length) {
+            changeCapacity(size);
+        }
+    }
+
+    private void changeCapacity(int requiredSize) {
+        int capacity = requiredSize + requiredSize / 2;
+        Object[] temp = new Object[capacity];
+        System.arraycopy(storage, 0, temp, 0, size);
+        storage = temp;
+    }
+
+    private void ensureCapacity() {
+        if (size == storage.length) {
+            changeCapacity(storage.length);
+        }
+    }
+
+    private void checkGetPosition(int index) {
+        if (index < 0 || index > size - 1) {
+            showOutOfBounds();
+        }
+    }
+
+    private void checkAddPosition(int index) {
+        if (index < 0 || index > size) {
+            showOutOfBounds();
+        }
+    }
+
+    private void showOutOfBounds() {
+        throw new ArrayListIndexOutOfBoundsException(OUT_OF_BOUNDS_MESSAGE);
+    }
+
+    private void removeBy(int index) {
+        Object[] temp = new Object[storage.length];
+        System.arraycopy(storage, 0, temp, 0, index);
+        System.arraycopy(storage, index + 1, temp, index, size - index - 1);
+        storage = temp;
+        size--;
     }
```

Comment:
```suggestion
    private void removeBy(int index) {
        System.arraycopy(storage, index + 1, storage, index, size - index - 1);
        size--;
    }
```
this should work

Reply:
Finally I understand how System.arraycopy works! 😀